Makefile

```diff
@@ -1,40 +1,120 @@
+
+	@# why some lints are disabled?
+	@#   gochecknoglobals - disabled because we do use quite a lot of globals
+	@#          goimports - executed above already, gofumports
+	@#              gofmt - ditto
+	@#                gci - ditto
+	@#                wsl - too pedantic about the formatting
+	@#             funlen - PENDING REFACTORING
+	@#           gocognit - PENDING REFACTORING
+	@#              godox - TODO
+	@#              gomnd - too many magic numbers, and too pedantic (even 2*x got flagged...)
+	@#        testpackage - several test packages still rely on private functions
+	@#             nestif - PENDING REFACTORING
+	@#           goerr113 - it mistaken pingcap/errors with standard errors
+	@#                lll - pingcap/errors may need to write a long line
+	@#       paralleltest - no need to run test parallel
+	@#           nlreturn - no need to ensure a new line before continue or return
+	@#   exhaustivestruct - Protobuf structs have hidden fields, like "XXX_NoUnkeyedLiteral"
+	@#         exhaustive - no need to check exhaustiveness of enum switch statements
+	@#              gosec - too many false positive
+	@#          errorlint - can't detect errors.Cause
+	@#      sqlclosecheck - the rows in dumpling is created in one function but closed in other functions
+	CGO_ENABLED=0 tools/bin/golangci-lint run --enable-all --deadline 120s \
+		--disable gochecknoglobals \
+		--disable goimports \
+		--disable gofmt \
+		--disable gci \
+		--disable wsl \
+		--disable funlen \
+		--disable gocognit \
+		--disable godox \
+		--disable gomnd \
+		--disable testpackage \
+		--disable nestif \
+		--disable goerr113 \
+		--disable lll \
+		--disable paralleltest \
+		--disable nlreturn \
+		--disable exhaustivestruct \
+		--disable exhaustive \
+		--disable godot \
+		--disable gosec \
+		--disable errorlint \
+		--disable sqlclosecheck \
+		$$($(PACKAGE_DIRECTORIES))
+	# pingcap/errors APIs are mixed with multiple patterns 'pkg/errors',
+	# 'juju/errors' and 'pingcap/parser'. To avoid confusion and mistake,
+	# we only allow a subset of APIs, that's "Normalize|Annotate|Trace|Cause".
+    # TODO: delete Errorf and New after we support standard code
+	@# TODO: allow more APIs when we need to support "workaound".
+	grep -Rn --exclude="*_test.go" -E "(\t| )errors\.[A-Z]" cmd v4 | \
+		grep -vE "Normalize|Annotate|Trace|Cause|Errorf|New" 2>&1 | $(CHECKER)
 
-bin/failpoint-ctl: go.mod
-	$(GO) build -o $@ github.com/pingcap/failpoint/failpoint-ctl
+lint: tools
+	@echo "linting"
+	CGO_ENABLED=0 tools/bin/revive -formatter friendly -config revive.toml $$($(PACKAGES))
 
-failpoint-enable: bin/failpoint-ctl
-# Converting gofail failpoints...
-	@$(FAILPOINT_ENABLE)
+tidy:
+	@echo "go mod tidy"
+	GO111MODULE=on go mod tidy
+	cd tools && GO111MODULE=on go mod tidy
+	git diff --exit-code go.mod go.sum tools/go.mod tools/go.sum
 
-failpoint-disable: bin/failpoint-ctl
-# Restoring gofail failpoints...
-	@$(FAILPOINT_DISABLE)
+.PHONY: tools
```

Comment:
.PHONY should include every target which is not on the filesystem...

Reply:
Yes. It sames that BR's makefile also need to be updated 🤣

Reply:
adding to .PHONY is not necessary (basically it just prevents a `./tools` file cause `make tools` do nothing) but at least keep the original ones 😓

Reply:
I didn't know the meaning of `.PHONY` before and directly copied this from BR. I have updated Makefile now.